Create client-side ordinations to avoid server-memory issues

grovduck opened the pull request:
This PR addresses #3 by leveraging [`sknnr`](https://github.com/lemma-osu/scikit-learn-knn-regression) to provide estimators for this package.  As noted in #3, running the estimators client-side will "interrupt" the GEE server-side flow and thus run training only once for all targets.  It also removes the duplication of implementing these ordination estimators in more than one repository.

This is still very much a work in progress and I will be addressing the following issues:
- [x] ~~Create a base class for all estimators to reduce duplicated code across estimators~~
- [x] Create client-side proxies of GEE feature collection objects so that the estimators can be run client-side
- [x] Following the design of `sknnr`, make `Raw` the base class
- [x] Introduce a derived class of `Raw` (currently called `Transformed`) to be the base class for all estimators that use transformations
- [x] Use `sknnr` transformation objects to handle client-side ordination
- [x] Move `predict` and `predict_fc` to `Transformed` and remove from derived classes
- [x] Create `@abstractmethod` methods in `Transformed` to handle transformations and implement in derived classes
- [x] Run tests
- [x] Cleanup (more tightening and class renaming)

As this PR turned out to be very ambitious, there are a few tasks that I decided not to tackle with this PR and turn into separate issues:

* Rework `Colocation` class so that user doesn't need to instantiate
* Test with other `n_components` and `spp_transform` values on `GNN` to ensure correctness
* Add `mode` keyword on feature collections to get distances and and tests
* Recreate spatial surfaces using `sknnr-spatial` and this training data to see if we can increase the number of expected matches of pixels in the test surfaces
* Fail quickly if `id_field` does not store an integer type

grovduck committed:
New base class (GeeKnnClassifier)

* Added method (__init__) and property (k_nearest) used across
  subclasses.

grovduck committed:
Add client-side proxies for Earth Engine objects

In order to run ordinations client-side, there needs to be a mechanism
for translating `ee.FeatureCollections` to python objects (in this case,
pydantic models).  Based on work from @aazuspan, these classes provide
multi-threaded functionality to convert feature collections in a
lower-memory context.  `Feature` and `FeatureCollection` have a few
helper functions to extract properties as values and lists.

grovduck commented:
@aazuspan, a stupid question.  I thought using:
```
from __future__ import annotations
```
would allow the use of `list` and `dict` (lowercase) as type hints (i.e., from [PEP 585](https://peps.python.org/pep-0585/#implementation)).  I'm not sure I remember how you got this to work in `sknnr`, but my tests are failing here and I think this is the reason.  I'm using this with `pydantic.BaseModel` derived classes like this:
```
from __future__ import annotations
from typing import Any
from pydantic import BaseModel


class Geometry(BaseModel):
    """Client-side proxy for ee.Geometry object"""

    type: str
    coordinates: list[Any]
```
Any help would be very much appreciated!

grovduck committed:
Change lists of env_columns and spp_columns to be client-side objects

aazuspan commented:
Interesting! If I'm remembering right, the `__future__` import works by effectively stringifying the type annotations to avoid the interpreter trying to evaluate incompatible ones. Looking at the failed test, I think `pydantic` is forcing those to be evaluated anyways, which is breaking the workaround. 

A little short on details, but pydantic/pydantic#2112 seems to be describing the same issue. On the opposite end of the spectrum, pydantic/pydantic#2678 has a *lot* of detail, although I haven't read deeply enough into that discussion to figure out what the conclusion was or whether it's relevant to backwards compatibility...

grovduck commented:
@aazuspan , thanks for your help finding those issues.  I'm not sure I understand much better than I did before other than it seems I'll have issues using `list` and `dict` in pydantic classes and trying to get 3.8 to pass.  If I go back to using `typing.List` and `typing.Dict`, I get all kinds of complaints from the `flake8-future-annotations` check.  Do you have any opinion on either:

1. dropping support for 3.8
2. using dataclasses instead of pydantic BaseModel
3. disabling the "FA" check

Typing is great to have, but kind of a pain as well!

aazuspan commented:
I'd be tempted to drop `3.8`... I know we considered that with `sknnr` and held off, but now we're 6 months past [Numpy's drop date](https://numpy.org/neps/nep-0029-deprecation_policy.html#drop-schedule), and it would sure simplify some things. With that said, option 3 would also be quick fix, so I wouldn't hesitate to do that if you have reservations about limiting Python version support.

> Typing is great to have, but kind of a pain as well!

💯

grovduck committed:
WIP: First cut at integrating sknnr estimators through client-side

This commit implements client-side fitting of four of the five
estimators (excluding Raw) and implements them using sknnr transformers
associated with each method.  For this commit only, we have introduced
a new method in each estimator called `train_client` whilst retaining
the original `train` which is a server-side train/fit.  The general
workflow is to convert the training data into a client-side feature
collection (class FeatureCollection), fit the model using the
transformer, and convert all needed data objects back into server-side
objects to work with `predict` and `predict_fc`.  For all but GNN,
`predict` and `predict_fc` are unchanged and tests are passing for both
client-side and server-side training.  For GNN, there were slight
modifications to `predict` and `predict_fc` in the transformations such
that only client-side fitting is working.  There are expected test
failures in this commit.

In addition, we have dropped support for python 3.8 in this commit as
types `list` and `dict` were not compatible with pydantic
BaseModel-derived classes for 3.8.

Still, much refactoring needs to be done to remove duplication across
the methods.

grovduck committed:
Fix sklean repository protocol

grovduck commented:
> I'd be tempted to drop `3.8`... I know we considered that with `sknnr` and held off, but now we're 6 months past [Numpy's drop date](https://numpy.org/neps/nep-0029-deprecation_policy.html#drop-schedule), and it would sure simplify some things

Great, thanks for your input.  I've dropped 3.8 (and added 3.12) to this commit.

grovduck committed:
Large refactor of most estimators into `Transformed` class

This commit changes the class hierachy to reflect that of `sknnr`.
The `GeeKnnClassifier` class has been renamed to `Raw`, a new abstract
class called `Transformed` derives from `Raw` and all other estimators
(`Eucliean`, `Mahalanobis`, `MSN`, and `GNN`) are all subclasses of
`Transformed`.  `Transformed` does most of the work in training and
predicting the estimators and four abstract methods delegate the
responsiblity of defining the transformer to the subclasses.

Additionally, `cca.py` and `ccora.py` are no longer needed as all
model fitting is now done through `sknnr`.

grovduck commented:
@aazuspan, if you have the time (and can stomach it), I'd love for you to take a quick look at the changes here.  The general thrust of this PR is to bring in the `sknnr` transformers [^1] into this package to do client-side fitting (training).  I didn't at all hew to the best-practices of nice, small, atomic commits so it's a bit to wade through, but the general pattern follows closely to what you've done with `sknnr`.

* `Raw` is the main base class and implements `train`, `predict` (spatial prediction), and `predict_fc` (feature collection [point] prediction).
* `Transformed` inherits from `Raw` but is an abstract class.  Its abstract methods set the protocol for doing the transformations which are delegated to the subclasses (`Euclidean`, `Mahalanobis`, `MSN`, and `GNN`).  It also implements `transform_image` and `transform_fc` which are called from `predict` and `predict_fc`, respectively.  Finally, it overrides `train` to do the fitting and the relevant objects are saved as EE server-side objects.

There are a few issues I know of at this point:

* Class names are bad (as they were originally in `sknnr`).  But given that the core neighbor finder is called `ee.Classifier.minimumDistance`, I felt a bit weird about calling these regressors again.  But maybe it's better to keep them consistent with `sknnr` rather than with `GEE`.
* I have a couple of no-op methods in `Raw` (`transform_image` and `transform_fc`) which just pass back the input argument.  Because I'm calling these methods from `Raw.predict` and `Raw.predict_fc`, mypy complained that these methods weren't originally defined in `Raw`.  I can't make those methods abstract without making all of `Raw` abstract, which would defeat the purpose.  Do you have a better way of defining these or changing `predict`/`predict_fc` so that it avoids this issue?  
* You'll see some weird values in the `test_model_spatial.py` file in `ESTIMATOR_PARAMETERS` (the last value in the tuple).  These represent the minimum number of expected matches between some check images that I created a while ago and the neighbor images created through the test, 360,000 being the maximum for a 600x600 image.  I honestly can't remember how I created those test images because `pynnmap` didn't have the capacity to run the different methods, so I'm guessing I did it with `yaImpute`.  But I think those check images might be worth recreating with `sknnr_spatial` to see if we can get higher values.

I know it's not in your nature, but please be brutally honest if you see some funky stuff and additional opportunities for refactoring.  The good thing is that we're currently passing tests!

[^1]: We only need the transformers as GEE's `ee.Classifier.minimumDistance` is reponsible for neighbor finding.  The transformers are used to set the ordination space.

grovduck commented on src/geeknn/ordination/_base.py:
There is no reason that we really need the `ordination` module any more.  I think it's probably better to move everything up a level.

aazuspan replied:
I was wondering but wasn't sure about future plans. Agreed!

grovduck commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+from joblib import Parallel, delayed
+from numpy.typing import NDArray
+from pydantic import BaseModel
+from sklearn.base import TransformerMixin
```
`scikit-learn` is currently not a dependency, but `sknnr` is.  If we know that `sknnr` requires `scikit-learn`, is it OK to leave out or best to be explicit in `pyproject.toml`?

aazuspan replied:
I would probably make it explicit even though we of course know `scikit-learn` will always be a dependency of `sknnr`, just to maintain best practices and avoid any confusion.

grovduck commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+        """Return specified properties as a 2-D numpy array."""
+        return np.array([f.get_properties(properties) for f in self.features])
+
+    @classmethod
+    def from_ee_feature_collection(
+        cls, fc: ee.FeatureCollection, num_threads: int = -1, chunk_size: int = 5000
+    ):
+        """Create a client-side FeatureCollection from an ee.FeatureCollection
+        using multiple threads and chunking to avoid memory issues.
+        """
+        size = fc.size().getInfo()
+        info = fc.limit(0).getInfo()
+        chunks = [fc.toList(chunk_size, i) for i in range(0, size, chunk_size)]
+
+        with Parallel(n_jobs=num_threads, backend="threading") as p:
+            chunk_data = p(delayed(chunk.getInfo)() for chunk in chunks)
+
+        info.update({"features": list(chain.from_iterable(chunk_data))})
+        return cls(**info)
```
This is more-or-less a direct lift of your `eeio` code.  I'm more than happy to give attribution, but just wasn't how best to do that if `eeio` remains a private repo.

aazuspan replied:
I appreciate it, but no need!

grovduck commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+
+    def transform_fc(self, fc: ee.FeatureCollection) -> ee.FeatureCollection:
+        """No-op placeholder."""
+        return fc
```
Not entirely satisfied with this design, but probably not too harmful or expensive?

grovduck replied:
Ditched this design in favor of suggestion from @aazuspan.

grovduck commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+
+    def train(
+        self, *, fc: ee.FeatureCollection, id_field: str, X_columns: list[str], **_
+    ):
```
How do you handle unused kwargs?  I want `y_columns` to be ignored if passed and I saw a post suggested this naming convention that any additional kwargs besides those listed should be ignored.

aazuspan replied:
I like this solution! I've just used `**kwargs` before and left it unused, but this is clearer and probably makes linters happier.

grovduck commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+            neighbor_fc = filter_neighbors(neighbor_fc, colocation_obj, self.id_field)
+
+        # Return the neighbors as a list of lists (plots x k)
+        return return_k_neighbors(neighbor_fc, self.k)
```
Probably a poor name for this function.  Maybe just `get_k_neighbors`?

aazuspan replied:
Yeah, I like `get_k_neighbors`.

grovduck commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+        return X, None
+
+    def train(
+        self, *, fc: ee.FeatureCollection, id_field: str, X_columns: list[str], **kwargs
```
Note to self: Whoops, used `**kwargs` here.

grovduck replied:
No, keep as `**kwargs` because `y_columns` can be passed to `GNN._get_modeling_X_y` and `MSN._get_modeling_X_y`.

aazuspan commented:
Absolutely, I'm excited to take a look and get a better idea of how this works! I tried running tests and it looks like I can access the table assets, but not the images. Do you mind sharing those? 

```python
FAILED tests/test_model_spatial.py::test_image_match[5-raw] - ee.ee_exception.EEException: Image.load: Image asset 'users/gregorma/gee-knn/test-check/raw_neighbors_600' not found (does not exist or caller does not have access).
FAILED tests/test_model_spatial.py::test_image_match[5-euc] - ee.ee_exception.EEException: Image.load: Image asset 'users/gregorma/gee-knn/test-check/euc_neighbors_600' not found (does not exist or caller does not have access).
FAILED tests/test_model_spatial.py::test_image_match[5-mah] - ee.ee_exception.EEException: Image.load: Image asset 'users/gregorma/gee-knn/test-check/mah_neighbors_600' not found (does not exist or caller does not have access).
FAILED tests/test_model_spatial.py::test_image_match[5-msn] - ee.ee_exception.EEException: Image.load: Image asset 'users/gregorma/gee-knn/test-check/msn_neighbors_600' not found (does not exist or caller does not have access).
FAILED tests/test_model_spatial.py::test_image_match[5-gnn] - ee.ee_exception.EEException: Image.load: Image asset 'users/gregorma/gee-knn/test-check/gnn_neighbors_600' not found (does not exist or caller does not have access).
```

grovduck commented:
> Do you mind sharing those?

Sorry about that.  Should be shared with your gmail account now.  Let me know if you have further issues.

grovduck commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+        neighbor_fc = transformed_fc.classify(
+            classifier=self.clf, outputName="neighbors"
+        )
+
+        # Add the IDs in from the original feature collection
+        ids = fc.aggregate_array(self.id_field)
+        neighbor_fc = crosswalk_to_ids(neighbor_fc, ids, self.id_field)
+
+        # Filter plots if colocation_obj is present
+        if colocation_obj is not None:
+            neighbor_fc = filter_neighbors(neighbor_fc, colocation_obj, self.id_field)
+
+        # Return the neighbors as a list of lists (plots x k)
+        return return_k_neighbors(neighbor_fc, self.k)
+
+
+class Transformed(Raw, ABC):
+    """Base class for estimators that require transforming the input data."""
+
+    @abstractmethod
+    def _get_transformer(self) -> TransformerMixin:
+        """Return the transformer to use for fitting."""
+        ...
+
+    @abstractmethod
+    def _get_X_means(self) -> list[float]:
+        """Return the means of the X columns."""
+        ...
+
+    @abstractmethod
+    def _get_X_stds(self) -> list[float]:
+        """Return the sample standard deviations of the X columns."""
+        ...
+
+    @abstractmethod
+    def _get_projector(self) -> list[list[float]]:
+        """Return the projection matrix."""
+        ...
```
In order to use a common pattern, some of the implementations of these methods are effectively no-op (e.g. the projector for `Euclidean` is the identity matrix, X_stds for `GNN` is all 1.0s), but I felt that it was a fairly small price to pay for the ease of coding the transformation all the same in `transform_image` and `transform_fc`.

grovduck replied:
For this commit, we will keep `_get_X_means`, `_get_X_stds`, and `_get_projector` implemented in the subclasses.  There is a bit of duplication between estimators, but not enough to see a default pattern emerge for each method.

aazuspan commented on tests/setup.py:
```diff
@@ -4,40 +4,39 @@
 
```
Was there an upside to using `setup.py` over `conftest.py`? I assume this is only getting run by `pytest` because it's imported by the other tests, so maybe there's a small chance the `ee.Initialize` could be missed if you ran a subset of future tests that didn't import from this module.

grovduck replied:
No upside at all! I frequently forget to use `conftest.py` and the nice utility of the automatic imports. I've renamed `setup.py` to `conftest.py` and move all test fixtures into that file.

grovduck commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+    def train(
+        self, *, fc: ee.FeatureCollection, id_field: str, X_columns: list[str], **kwargs
+    ):
+        """Train the estimator, which consists of many steps:
+
+        1. Convert the server-side ee.FeatureCollection to a client-side
+           FeatureCollection
+
+        2. Get the input arrays for modeling from the feature collection
+
+        3. Get the transformer for this estimator and tranform the X data
+
+        4. Create server-side objects of needed data for prediction, including
+           setting the minimumDistance classifier
+        """
```
This level of detail is left as fairly explicit for reviewers - is it needed as a docstring?  Better as comments?

aazuspan replied:
Good question! I like the detail (it was very helpful for me), but I can see not wanting to distract users with implementation info. The [Numpy docstring style guide](https://numpydoc.readthedocs.io/en/latest/format.html#notes) suggests you can discuss algorithms in the `Notes` section, so maybe that's a good compromise?

grovduck replied:
Moved to a `Notes` section, but docstrings will be part of another issue/PR

grovduck commented on src/geeknn/ordination/gnn.py:
```diff
@@ -1,218 +1,38 @@
-        id_field = self.id_field
+    def _get_transformer(self):
+        # return CCATransformer(self.n_components)
+        return CCATransformer(n_components=self.num_cca_axes)
```
Still need to ensure that `n_components` is working as intended.  I don't have any explicit tests for this as of now.

grovduck replied:
Punting on this one for now.  This should be captured as a separate issue that tests for multiple `n_components` values.

aazuspan commented on tests/test_model_spatial.py:
```diff
@@ -20,28 +20,28 @@ def training_data():
 
 @pytest.fixture()
-def env_image():
+def X_image():
```
Just to simplify things a little bit, you could rename `get_covariate_image` to `X_image`, turn it into a fixture, and remove this function, right? I think the same applies to the other fixtures that are based on a getter in `setup.py`, unless some of them need to be module-scope.

grovduck replied:
Yep, makes total sense. I've done this with the other getters as well (`get_training_data` and `observed_ids`).

aazuspan commented:
All tests passing now, thanks!

aazuspan commented on tests/setup.py:
```diff
@@ -49,12 +48,12 @@ def join_feature(f):
     # ordering of the bands in the stacked environmental image
     unwanted = ee.List([id_field, "system:index"])
-    spp_columns = spp.first().propertyNames().removeAll(unwanted).sort()
+    y_columns = y.first().propertyNames().removeAll(unwanted).sort().getInfo()
```
Out of curiosity, why define the X cols as a constant and calculate the y cols?

grovduck replied:
Good question and maybe there is a more clever way of doing this (potentially when selecting the bands from the `X_image`)? When I do:
```
unwanted = ee.List([id_field, "system:index"])
X_columns = X.first().propertyNames().removeAll(unwanted)
```
the columns aren't sorted in the same order as the bands in the `X_image`. The band names in the GEE assets are actually ["b0", "b1", etc.], so I think I was taking the easy way out and just assigning names (e.g. `.rename(X_COLUMNS)`) rather than going to the "hard" work of correcting band names on the asset and then using `.select(X_COLUMNS)`. Then the feature collection property names could be in any order and it should work (or possibly raise an error if all names weren't found). Probably worth the latter approach?

aazuspan replied:
That seems totally fair, and I'm not sure it's worth the extra effort to avoid renaming, since this is just test code after all. I guess I was mostly thinking about this from the opposite direction: could you define `y_columns` as a constant to avoid the `getInfo` call during testing? It's probably a non-issue though since the call is only made once and is likely very fast.

grovduck replied:
Ah, that makes sense (about `y_columns` being a constant as well).  I think I've typically shied away from this as our species matrices have been lots of columns and I was too lazy to list them all out.  (Obviously it's a cut-and-paste job if I do it once.)  I hadn't really considered the overhead of the `.getInfo` call.  I think I'll leave it for now (the whole testing suite is a bit brittle) and see if other patterns emerge if I bring in other datasets.

aazuspan replied:
Not wanting to include a giant list of columns in the code is fair! Sounds like a good plan.

aazuspan commented on src/geeknn/ordination/utils.py:
```diff
@@ -1,17 +1,24 @@
 from pydantic import BaseModel, ConfigDict
 
 
 class Colocation(BaseModel):
```
From a user perspective, I don't love the idea of having to instantiate and track a `Colocation` object. The implementation looks nice and clean, but as a user API it adds some complexity to the typical prediction workflow, if I'm understanding it right. A couple alternative ideas you could consider:

1. Make the `Colocation` class internal and add an optional `location_field: str` param to `predict_fc`, which could be passed to `filter_neighbors` and used internally to identify the colocated features and remove them.
2. Remove the `Colocation` class and have `utils.get_colocation_fc` return a `FeatureCollection` with the `location_field` stored as a property that could be accessed by `filter_neighbors`. The user would still need to keep track of it, but wouldn't need to worry about an unfamiliar class. The downside is that it would be easier to accidentally modify the colocation FC or pass in the wrong one.

There may be some limitations to these ideas I'm not considering, or maybe they don't work with your future plans, so take them with a grain of salt!

grovduck replied:
I definitely understand what you're saying here, but I'm struggling with one aspect of this.  Setting up this crosswalk requires the user to specify the feature collection that holds the crosswalk (as well as the fields involved).  Where would the user specify the feature collection if this class was internal?  I think I'm totally missing something.

aazuspan replied:
I'm still pretty fuzzy on this part of the workflow, so any confusion is definitely coming from me! I had assumed that the typical usage would be to have one feature collection with all of the species, environmental, plot ID, and optional location ID fields, and you would need to feed that into `Colocation` to get a new `colocation_obj`, then feed both into `predict`, which seemed slightly redundant. After taking a closer look at the tests, I see that you're bringing in a separate collection for the plot/location ID crosswalk. I take it that's a more typical workflow?

Would it be possible/preferable to require the user to pre-join the location ID field into the main `fc` if they want to filter co-located plots? It would add another step up-front (maybe there could be a helper func, since joins in EE are admittedly a hassle), but overall maybe it would simplify things by hiding the colocation details and forcing the user to just deal with one FC and field names?

You have a much better idea of the typical workflow than I do, so take any of my input here with a grain of salt!

grovduck replied:
> I had assumed that the typical usage would be to have one feature collection with all of the species, environmental, plot ID, and optional location ID fields, and you would need to feed that into `Colocation` to get a new `colocation_obj`, then feed both into `predict`, which seemed slightly redundant. After taking a closer look at the tests, I see that you're bringing in a separate collection for the plot/location ID crosswalk. I take it that's a more typical workflow?

Ah, that makes a lot more sense and it was likely my bad explanation or lack of details that led to the confusion.  If we step back to stock GNN mapping for NWFP, I have a single file that crosswalks plot IDs to location IDs and this file can be used across all modeling regions so that we don't actually add the location ID to the files we use in modeling.  So, again because of historical patterns, I've always imagined this to be a separate file that isn't specifically tied to modeling.

> Would it be possible/preferable to require the user to pre-join the location ID field into the main fc if they want to filter co-located plots? It would add another step up-front (maybe there could be a helper func, since joins in EE are admittedly a hassle), but overall maybe it would simplify things by hiding the colocation details and forcing the user to just deal with one FC and field names?

This seems like a reasonable way to handle it.  These types of decisions will come into play for `sknnr` (or the wrapper package around `sknnr` that works with AA) as well.  This would still allow the user to have a main lookup feature collection, but not have the user be responsible for instantiating the object.  

Having said this and in the interest of time, I think this one feels like it requires a bit more thought on how to best implement this, so I'm going to punt on this and create a new issue to deal with it.  So we'll still have the user-managed `Colocation` class in this PR, but hopefully remedied soon.

aazuspan replied:
Sounds like a good plan!

aazuspan commented on src/geeknn/ordination/utils.py:
```diff
@@ -1,17 +1,24 @@
 class Colocation(BaseModel):
+    """Class for identifying colocated features in a feature collection"""
+
     model_config = ConfigDict(arbitrary_types_allowed=True)
```
For my own understanding, can you explain why this was needed? I'm a `pydantic` newb!

grovduck replied:
My understanding is somewhat lacking on this, but I believe the config is needed for "arbitrary" classes that pydantic doesn't know anything about. These would be user-defined classes (e.g. in [this example](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.arbitrary_types_allowed)) or other non built-in classes (e.g., numpy arrays, pandas dataframes). Here I'm using it for `ee.FeatureCollection`.

aazuspan replied:
Interesting, thanks for the reference! I definitely have some more reading to do.

aazuspan commented on tests/test_model_spatial.py:
```diff
@@ -20,28 +20,28 @@ def training_data():
-    nn = model.predict(env_image=env_image, mode="CLASSIFICATION").retile(32)
-    diff_nn = nn.subtract(check_img).abs()
-    frequency = diff_nn.reduceRegion(
+    nn_image = model.predict(X_image=X_image, mode="CLASSIFICATION").retile(32)
```
Ooh, an undocumented `ee.Image` method? Interesting! Does `retile` work together with or replace `tileScale` somehow?

grovduck replied:
Yeah, I'm pretty sure I'm not supposed to be advertising this one (or maybe even using it). This was something they put in the API a while back, but never really documented it (nor took it away!). AFAIK, `tileScale` can only be used with reducing or sampling(?), but I think it's the same concept. I believe the way Noel explained it to me was that `retile` just makes the map tiles smaller so that you're passing less data and, therefore, less likely to run into memory issues. In this case, if I don't use it, tests don't pass and I get `User memory limit exceeded` errors. If we were doing exports, I believe this wouldn't be an issue (different memory limits, correct?).

But I think the inclusion of `tileScale` in the `reduceRegion` call is now not needed because we have this `.retile` call in there.

aazuspan replied:
Very cool, thanks for the inside scoop! I won't spread the word, but I am curious to try it out. I wonder why it didn't make it into the public methods, since it seems like a more flexible approach than `tileScale`.

aazuspan commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+            chunk_data = p(delayed(chunk.getInfo)() for chunk in chunks)
+
+        info.update({"features": list(chain.from_iterable(chunk_data))})
+        return cls(**info)
```
The way this works together with `pydantic` to auto-generate the `Feature` objects is awesome! If I ever get back to `eeio`, I think I'll need to snag this!

grovduck replied:
Well, it was pretty much all your work (other than the `pydantic` fluff)! Have you played around at all with [`FastAPI`](https://fastapi.tiangolo.com/)? Very cool and built on top of `pydantic` models.

aazuspan replied:
I haven't, but I've heard great things! It's definitely at the top of my list next time I need a Python server.

aazuspan commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+    columns: dict[str, Any]
+    version: int
+    id: str
+    properties: dict[str, Any]
```
I'm not sure how much effort you want to put into making these work for every case, but I think having a `version`, `id`, and `properties` is only guaranteed for ingested assets, so a generated FC (e.g. created with `ee.Feature.randomPoints`) would fail to validate.

grovduck replied:
Ooh, good call! I didn't realize that. I set `Optional[]` on these fields (as well as on the geometry field of Feature as that is allowed to be a null geometry).

aazuspan commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+class Raw:
+    def __init__(self, k=1, max_duplicates=None):
+        self.k = k
+        self.max_duplicates = max_duplicates if max_duplicates is not None else 5
```
```suggestion
    def __init__(self, k=1, max_duplicates=5):
        self.k = k
        self.max_duplicates = max_duplicates
```

Was there a reason to avoid this?

grovduck replied:
I think this was a carry-over from the original port from JS where I couldn't provide default parameters. Thanks for catching this!

aazuspan commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+        """
+        return self.k + self.max_duplicates
+
+    def get_ids(self, fc: FeatureCollection, id_field: str) -> list[Any]:
```
Seems slightly odd to have as an instance method. Maybe a `staticmethod` or a `utils` function?

grovduck replied:
You're absolutely right. I don't think the function is needed at all. I've just inlined this functionality in `Transformed.train` to be:
```
self.fc = scores_to_fc(
    client_fc.aggregate_array(id_field).tolist(),
    X_transformed.tolist(),
    id_field,
)
```

aazuspan commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+
+    def get_ids(self, fc: FeatureCollection, id_field: str) -> list[Any]:
+        """Return the IDs from the feature collection."""
+        return list(map(int, fc.aggregate_array(id_field)))
```
Will IDs always be castable to `int`, or could they be non-numeric?

grovduck replied:
Well, they *should* be ints 😄. But your point is well-taken ... Because these IDs are eventually used in an `ee.Image` context (i.e., the neighbor images), things will go wrong if they are not integers. Now that I've deleted the `get_ids` function, I think `utils.scores_to_fc` would actually work (ignoring the type hint of `ids` as `list[int]`), so the error wouldn't actually show up until running `predict`.

Do you think I need to fail quickly here if the IDs are not integers?

aazuspan replied:
Ah, I see! 

> Do you think I need to fail quickly here if the IDs are not integers?

I think it might be worth it, mostly for the benefit of being able to throw a more helpful error than EE probably would. I suppose there's no way to easily validate that field using `pydantic`, since the values for `Feature.properties` need to be `Any` to accommodate other fields, right?

grovduck replied:
> I suppose there's no way to easily validate that field using `pydantic`, since the values for `Feature.properties` need to be `Any` to accommodate other fields, right?

Exactly right.  And I think it's complicated by the fact that I'm getting IDs both client-side (through `FeatureCollection.aggregate_array`) and server-side (through `ee.FeatureCollection.aggregate_array`).  I'd like to use the same logic to validate both, but that will likely require a `.getInfo()`.  I can probably make this a bit faster by assuming the first item is representative of the list and use `.limit(1)` to just retrieve a single record from the server-side collection, but it feels like I'll need to use `isinstance` for type checking the client-side object.

I've added this to the list of issues that I'm not going to address with this PR and will deal with with subsequent (smaller) issues/PRs.

aazuspan commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+
+    def predict(self, X_image: ee.Image, mode: str = "CLASSIFICATION"):
+        """Predict the nearest neighbors for the given covariate (X) image."""
+        self.clf = self.clf.setOutputMode(ee.String(mode))
```
Maybe create a copy of `self.clf` instead of replacing it, to avoid permanently setting the output mode? Otherwise this would affect `predict_fc`, right?

grovduck replied:
Thanks. That one was a silly error ...

aazuspan replied:
Gotcha!

This may reveal my lack of understanding of how this all fits together, but I'm curious why there's no `mode` option for `predict_fc`? Does regression make sense in an image context but not in a feature context for some reason?

grovduck replied:
> This may reveal my lack of understanding of how this all fits together

As always, your understanding is spot on.  I just haven't yet done anything with `REGRESSION` mode with feature collections.  For me, `REGRESSION` is a bit of a misnomer for `ee.Classifier.minimumDistance` as it's just returning the distances from each of the neighbors.  `sknnr` is way further ahead in this manner right now in that it (optionally) uses the NN distances in prediction of attributes.  For now, `geeknn` is mostly using just the neighbors in prediction and typically using equal weighting or our default k=7 weighting and those are separate modules, currently only in the JS version of this package.  For immediate needs, I just need to return neighbor IDs, but I think it makes sense to implement attribute prediction, at which point we will need to call `predict` in both `CLASSIFICATION` and `REGRESSION` modes to get the needed components for both images and feature collections.  I have a note to myself to create an issue to address this.

aazuspan commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+        return np.array([f.get_properties(properties) for f in self.features])
+
+    @classmethod
+    def from_ee_feature_collection(
```
It would be nice to cache the slow `getInfo` calls in cases where you want to train a model a few times with different parameters but the same dataset. Just putting an `lru_cache` on this method would be slightly weird in that the cache would be tied to the `num_threads` and `chunk_size` args, but I can't really think of a good alternative and there's probably no practical downside to that...

grovduck replied:
I haven't used `@lru_cache` too often.  Is there a `maxsize` you'd recommend for this?
```
    @classmethod
    @lru_cache(maxsize=1)  # default is 128
    def from_ee_feature_collection(
        cls, fc: ee.FeatureCollection, num_threads: int = -1, chunk_size: int = 5000
    ):
        """Create a client-side FeatureCollection from an ee.FeatureCollection
        using multiple threads and chunking to avoid memory issues.
        """
        ...
```

aazuspan replied:
Great question. I don't know of any good rules of thumb for this, so I think I'd probably just have to make an educated guess. The default 128 seems as little risky in the case where someone is repeatedly grabbing large, slightly different collections, so maybe something in the range of 32 or 64 would make sense? I imagine that would cover most typical use cases where there would only ever be a handful of different collections in the cache.

grovduck replied:
Excellent, I'll start with 32.

aazuspan commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+        input_properties: ee.List,
+    ) -> None:
+        """Set the EE minimumDistance classifier for this estimator."""
+        self.clf = ee.Classifier.minimumDistance(
```
Just a thought with no strong opinion either way, but what do you think about setting `self.clf` in `__init__` and replacing it with the trained version in `train`? That would simplify things slightly, but maybe there are downsides I'm not considering.

grovduck replied:
Yep, good idea.  I've done that in a recent commit and renamed `Raw._get_classifier` to `Raw._train_classifier` since that's it's only responsibility now.  Because I'm still calling it from both `Raw.train` and `Transformed.train` with different signatures, I think I still need this method.

aazuspan replied:
Ah, I missed that it was being used by both `train` methods, so keeping it in a separate `_train_classifier` method looks like a great solution!

aazuspan commented on tests/test_model_spatial.py:
```diff
@@ -20,28 +20,28 @@ def training_data():
 
-def run_method(kls, options, training_data, env_image, check_img):
+def run_method(kls, options, training_data, X_image, check_image):
     """Run predict on the given estimator, difference it against a
     reference images, and return the frequency of zero differences"""
```
Very clever approach for testing results!

aazuspan commented on src/geeknn/ordination/gnn.py:
```diff
@@ -1,218 +1,38 @@
+    SPP_TRANSFORM_FUNC = {
+        "SQRT": lambda x: np.sqrt(x),
+        "LOG": lambda x: np.log(x),
+        "NONE": lambda x: x,
```
Maybe allow `spp_transform=None` instead of using `"NONE"` as a key?

grovduck replied:
I think I'm being dense here, but I'm trying to figure out how to make this change, but are you suggesting something like this?
```
class GNN(Transformed):
    Y_TRANSFORM_FUNC = {
        "SQRT": lambda x: np.sqrt(x),
        "LOG": lambda x: np.log(x),
    }

    ...

    def _get_modeling_X_y(self, *, client_fc, X_columns, y_columns):
        X = client_fc.properties_to_array(X_columns)
        y = client_fc.properties_to_array(y_columns)
        transform_func = self.Y_TRANSFORM_FUNC.get(self.y_transform.upper(), lambda x: x)
        y = transform_func(y)
```
I don't think this is what you meant exactly ...

aazuspan replied:
I think what I had in mind (but didn't explain well) was using `None` as the key instead of the string `"NONE"`, so the user could specify `y_transform=None` to apply the no-op transform (I had to check to confirm, but `None` is a valid dict key). Returning the no-op as a default from `get` is a little more concise though. My main goal was just to remove a string option since I find them hard to remember and a little messy to validate.

grovduck replied:
Now I understand, thanks for clarifying.  I'm going with your advice to make this directly a callable.  Much cleaner and no validation on the strings as you say.

aazuspan commented on src/geeknn/ordination/gnn.py:
```diff
@@ -1,218 +1,38 @@
+
     def __init__(self, k=1, spp_transform="SQRT", num_cca_axes=8, max_duplicates=None):
-        self.k = k
         self.spp_transform = spp_transform
```
Fail fast with invalid params by getting and storing the `spp_transform` func during init? 

Another option you could consider would be just taking a callable, e.g. `spp_transform=np.sqrt` - more flexible if users want to pass a custom callable, but a little less convenient if they have to import `numpy` to get `np.log`.

grovduck replied:
Thanks! I went with this approach.

aazuspan commented on src/geeknn/ordination/gnn.py:
```diff
@@ -1,218 +1,38 @@
-        zipped = neighbor_fc.toList(neighbor_fc.size()).zip(ids)
-        id_field = self.id_field
+    def _get_transformer(self):
+        # return CCATransformer(self.n_components)
```
Leftover comment?

grovduck replied:
Yep, this was associated with the fact that I hadn't yet fully tested `n_components` (and still haven't).  I've deleted these comments in `MSN` and `GNN`.

aazuspan commented on src/geeknn/ordination/euclidean.py:
```diff
@@ -1,131 +1,20 @@
-        fc = ee.FeatureCollection(fc)
-        self.id_field = ee.String(id_field)
-        self.env_columns = ee.List(env_columns)
+    def _get_X_means(self) -> list[float]:
+        return self.transformer_.mean_.tolist()
```
Since the same `_get_X_means` and `_get_X_stds` implementation is used by `Euclidean`, `MSN`, and `Mahalanobis`, do you think it's worth implementing in the base class and overriding for `GNN`? The same could probably be done with `_get_projector` and overriding for `Euclidean`. 

I'm not 100% sure I like that suggestion, but something to consider.

grovduck replied:
Actually, `Euclidean._get_X_means` is very slightly different from `Mahalonobix._get_X_means` (and `MSN._get_X_means`) in that the latter two get the information from `self.transformer_.scaler_.mean_` and `Euclidean` gets it directly from `self.transformer_.mean_`.  Maybe something to think about changing in `sknnr` itself to provide a common interface to these data?  (`._get_projector` has similarly subtle differences).

For now, I'll keep the implementations in the derived classes even if there is a bit of duplication.

aazuspan replied:
Ah, I should've looked more carefully! Thanks for pointing it out.

As far as providing a common interface in `sknnr`, would that mean adding a `Mahalanobis.transformer_.mean_` property that points to the `transformer_.scaler_.mean_`? I haven't looked back at the code or thought through the implications, but I'm open to that if you think it's worthwhile. Certainly not necessary just to reduce the minor duplication here, though!

grovduck replied:
> I haven't looked back at the code or thought through the implications, but I'm open to that if you think it's worthwhile.

I'm not sure it *is* actually worth it.  I think this package might be one of the only places where we need access to the innards of the transformers as we need to convert to server-side objects.  I'm guessing that users of both `sknnr` and `geeknn` will mostly be using the estimators over the transformers.  I may be wrong about this, but I don't think it's worth the change right now.

aazuspan replied:
> I'm guessing that users of both sknnr and geeknn will mostly be using the estimators over the transformers. I may be wrong about this, but I don't think it's worth the change right now.

Yeah, I would assume the same thing. Even for people using the transformers directly, I'm guessing that accessing those fit attributes will be a corner case. Leaving this for now sounds good to me :)

aazuspan commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+        X_image = X_image.select(self.X_columns)
+
+        # Transform image
+        X_image = self.transform_image(X_image)
```
> Do you have a better way of defining these or changing predict/predict_fc so that it avoids this issue?

I do have one idea for you to consider: remove this line and the corresponding line in `predict_fc`, then override those methods in `Transformed` to just apply the transform and pass the transformed Image/FC to this superclass. So `Transformed.predict` would look something like:

```python
def predict(self, X_image, mode: str = "CLASSIFICATION"):
    transformed_image = self.transform_image(X_image)
    super().predict(transformed_image, mode=mode)
```

It adds a *little* complexity, but avoids the placeholder methods and the strangeness of calling a transform method on a raw "regressor" that doesn't apply any transform.

grovduck replied:
Great idea!  With this latest commit, I've incorporated both this idea and your other idea for using `@singledispatchmethod` to make `Raw` and `Transformed` a bit more streamlined.  In order to make this refactoring, I think I have to create private methods (`Raw._predict_image` and `Raw._predict_fc`) as the setup before each of these calls is a bit different when using a `Raw` object versus a `Transformed` object.  For example, I created a new method (`Raw._get_ordered_X_image`) that passes the `X_image` and returns an image with possible reordering of the bands based on `self.X_columns`.  I couldn't put that into `Raw.predict` because a transformed image no longer has the same band names as the original (the same type of issues that we were dealing with when doing `sknnr` transformations).  

But your fix is a much better flow and more true to what I think of as good OOP practice as well as getting rid of the no-op functions!  Thanks for the good suggestion.

aazuspan replied:
>  In order to make this refactoring, I think I have to create private methods (Raw._predict_image and Raw._predict_fc) as the setup before each of these calls is a bit different when using a Raw object versus a Transformed object.

Ah, tricky. This looks like a good solution considering the constraints!

aazuspan commented on src/geeknn/ordination/_base.py:
```diff
@@ -0,0 +1,282 @@
+            .arrayFlatten([band_names])
+        )
+
+    def predict_fc(self, fc: ee.FeatureCollection, colocation_obj=None):
```
I've gone back and forth on whether or not this is a terrible suggestion, but you could consider using a [single dispatch](https://docs.python.org/3/library/functools.html#functools.singledispatch) for the `predict` and `transform` methods, so they would accept either an `Image` or `FeatureCollection` and internally dispatch to the correct method. It makes the public API a little simpler, but possibly also more confusing...

grovduck replied:
This is a really cool idea - I've never dived deeply enough into `functools` to know it was there (takes me back to my dinosaur C days where varying the arguments creates different functions).  It looks like it's a pretty simple fix if we have an initial argument with the rest being keyword arguments:
```
from functools import singledispatchmethod

class Raw:
    @singledispatchmethod
    def predict(self, arg, **kwargs):
        raise NotImplementedError

    @predict.register
    def _(self, X_image: ee.Image, mode: str = "CLASSIFICATION"):
        ...

    @predict.register
    def _(self, fc: ee.FeatureCollection, colocation_obj=None):
        ...

fc_prediction = model.predict(fc, colocation_obj=colocation_obj)
img_prediction = model.predict(X_image, mode="CLASSIFICATION")
```
But then it doesn't look like you can call them like this anymore:
```
fc_prediction = model.predict(fc=fc, colocation_obj=colocation_obj)
img_prediction = model.predict(X_image=X_image, mode="CLASSIFICATION")
```
From your perspective, is the first way a more natural way to call this?  What were your reservations about suggesting this initially?

aazuspan replied:
> takes me back to my dinosaur C days where varying the arguments creates different functions

Exactly, function overloads in Python! (if you're feeling really nostalgic, have you seen [pointers.py](https://github.com/ZeroIntensity/pointers.py)?)

> From your perspective, is the first way a more natural way to call this?

Yeah, I think in a function like this where the first argument is the focus of the function and the rest of the args just modify the operation, my instinct is usually to make it positional.

>  What were your reservations about suggesting this initially?

My main concern is that the relevant args depend on the input type, which makes for a confusing docstring and signature. I think the only time I've seen single dispatch in use is in the Planetary Computer SDK where there's a [sign method](https://github.com/microsoft/planetary-computer-sdk-for-python/blob/30cc12bebe8a10bdd99bb0b32956cc2c8078a58d/planetary_computer/sas.py#L85) that dispatches for a bunch of different object types, but they all take the same args, so they avoid that complexity.

grovduck replied:
> (if you're feeling really nostalgic, have you seen [pointers.py](https://github.com/ZeroIntensity/pointers.py)?)

OMG, that's hilarious (and frightening!)

> My main concern is that the relevant args depend on the input type, which makes for a confusing docstring and signature.

Definitely a warning to be heeded!  

As I noted above, I went ahead and implemented `@singledispatchmethod`s for `Raw.predict`, the overloaded `Transformed.predict`, and `Transformed.transform`.  I did take an initial stab at a docstring for both `Raw.predict` and `Transformed.transform`, which I leaned on ChatGPT to help create and get the expected formatting.  

I was hoping to reuse the docstring from `Raw.predict` for `Transformed.predict` with the use of `functools.wraps`, but I couldn't quite get it to work.  The dispatching doesn't seem to work nicely with `@wraps`.  For now, I don't have a docstring, but I'm guessing I should just copy it from `Raw.predict`?

grovduck replied:
For now, I've copied the docstring directly from `Raw.predict` to `Transformed.predict`.

aazuspan reviewed (commented):
There's a lot of really cool stuff going on here! I'm still trying to grok how it all fits together, so let me know if any of my comments are totally off base, but I made a first pass at least.

> I felt a bit weird about calling these regressors again. But maybe it's better to keep them consistent with sknnr rather than with GEE.

Choosing names remains the hardest part of programming! I'm not sure that staying consistent with `sknnr` is necessarily the perfect choice, but it does seem like a defensible one at least, and would make it easy for someone familiar with one package to adapt to the other (although if you ever tried using them in a single script there might be some namespace headaches...). I think that's where I would lean by default, just because I don't have any better ideas.

grovduck commented:
> but I made a first pass at least

@aazuspan, thanks for such a thorough review.  I'll be picking through it over the next couple of days (then back to `sknnr`!)

grovduck committed:
Rename `setup.py` to `conftest.py` and move test fixtures into it.

grovduck committed:
Remove `tileScale` from `test_model_spatial::run_method`.

grovduck committed:
Fix client-side objects to accept null fields for some properties.

grovduck reviewed (commented):
@aazuspan, I think I had started a review myself, but had never submitted it.  Hopefully you can see my responses to your comments.

grovduck committed:
Change argument signature of GNN.__init__ and rename attributes

grovduck committed:
Remove `get_ids` methods and inline the functionality where it's called

grovduck committed:
Make `clf` local in `Raw.predict` rather than overwriting `self`

aazuspan commented:
> Hopefully you can see my responses to your comments.

I do see a couple, but I got an email notification about a bunch of comments that aren't showing up for me, e.g. your answer about `retile` or the int IDs... Not quite sure what's going on there.

grovduck commented:
> I do see a couple, but I got an email notification about a bunch of comments that aren't showing up for me, e.g. your answer about `retile` or the int IDs... Not quite sure what's going on there.

I really messed everything up.  But I think here is what happened (chronologically):
1. I started a review right after I initiated the PR but didn't submit it.
2. You provided your review.
3. I started responding to your comments, but they were in the context of my ongoing review.
4. I finally realized that I hadn't yet submitted my initial review.
5. All my responses to you got duplicated in my review and had no context (e.g., your original comment) in my review
6. I deleted all my duplicated comments, making sure that they stayed when in response to your comments.
7. Upon refresh, all my responses disappeared.

So, I'll try to go back and put in what I had written (I don't think deleted comments can be retrieved), but probably best to use the emails that were sent as the definitive source.  I think the crux is that one shouldn't respond when an active review is ongoing.  Sorry about this!

grovduck committed:
Change `Geometry | None` to `Optional[Geometry]` for pydantic

aazuspan commented:
Oh no, that's a hassle to have to re-enter all your responses. I think I have the text of all your responses in the notification emails, I just don't necessarily know what they were responses *to* (although I can probably guess in most cases). Do you have those, or would it be helpful if I forwarded the email?

Github is a pretty smooth experience overall, but figuring out where/if review comments are going to show up seems like it's always a little bit of a gamble, so you're not alone!

grovduck commented:
> I think I have the text of all your responses in the notification emails, I just don't necessarily know what they were responses _to_ (although I can probably guess in most cases). Do you have those, or would it be helpful if I forwarded the email?

If you have those and it would be easy enough to send, that would be great.  I don't get a copy of changes I make.

grovduck committed:
Change all instances of `|` to `Optional[]` for pydantic

grovduck committed:
Raw and Transformed class refactoring based on comments

- Remove no-op `Raw.transform_image` and `Raw.transform_fc`
- Convert `Raw.predict`, `Transformed.predict`, and
  `Transformed.transform` into `singledispatchmethod`s so that image and
  feature collection transformation and prediction comes from
  `transform` and `predict`, respectively.

grovduck committed:
Add `scikit-learn` as a dependency`

grovduck committed:
Rename `return_k_neighbors` to `get_k_neighbors`

grovduck committed:
Comments and docstrings cleanup

grovduck committed:
Initialize minimumDistance classifier in __init__ with later training

grovduck commented:
> Choosing names remains the hardest part of programming! I'm not sure that staying consistent with `sknnr` is necessarily the perfect choice, but it does seem like a defensible one at least, and would make it easy for someone familiar with one package to adapt to the other (although if you ever tried using them in a single script there might be some namespace headaches...). I think that's where I would lean by default, just because I don't have any better ideas.

I know it's not what you suggested, but I'm considering the following name changes:
- `Raw` -> `RawKNNClassifier`
- `Euclidean` -> `EuclideanKNNClassifier`
- `Mahalanobis` -> `MahalanobisKNNClassifier`
- `MSN` -> `MSNClassifier`
- `GNN` -> `GNNClassifier`

I'm thinking of this from the following perspective:
- With `sknnr`, we chose our naming off the main estimator we were deriving - `KNeighborsRegressors`.  With this package, I'm worried if we call them regressors, users will be more confused by that terminology rather than the more familiar `Classifier` terminology.  The term isn't completely accurate, but GEE calls their estimators classifiers even if run in regression mode.
- If we ever did use the two packages together (something like speed testing), we'd avoid the naming conflicts that you mention.

Thoughts?

aazuspan commented:
>The term isn't completely accurate, but GEE calls their estimators classifiers even if run in regression mode.

Good point!

I think your logic makes sense, and I'm happy to go with the `Classifier` naming scheme. As you suggested, that should be a more familiar interface for users with some Earth Engine ML experience, who are probably the most likely to find and use this.

grovduck committed:
Add lru_cache on `FeatureCollection.from_ee_feature_collection`

grovduck committed:
Change `y_transform` to take `Callable` instead of `str`

grovduck committed:
Rename all estimator to specify that they are NN classifiers

grovduck committed:
Move all modules out of the `ordination` subpackage and into the root

grovduck committed:
Copy identical docstring from `RawKNNClassifier.predict` to `TransformedKNNClassifier.predict`

grovduck commented:
@aazuspan, you've been incredibly gracious with your time on reviewing this PR.  Thank you so much.  I'm running out of steam a bit and will migrate these issues you've identified below into separate issues.

- Rework `Colocation` class so that user doesn't need to instantiate
- Test with other `n_components` and `spp_transform` values on `GNN` to ensure correctness
- Add `mode` keyword on feature collections to get distances and and tests
- Recreate spatial surfaces using `sknnr-spatial` and this training data to see if we can increase the number of expected matches of pixels in the test surfaces
- Fail quickly if `id_field` does not store an integer type

There is still a bit of work to do, but I'd like to merge in this PR now, mostly because I need to test with the larger SERVIR workflow this coming week.  I've learned that this was much too ambitious of a PR on its own!

aazuspan commented:
Sounds great @grovduck! Thanks for helping me get up to speed on this. I'm excited to start exploring it with some real data!

grovduck commented:
> Sounds great @grovduck! Thanks for helping me get up to speed on this. I'm excited to start exploring it with some real data!

@aazuspan, even though there are still many things to do on this, I wouldn't have gotten nearly as far without your helpful reviews and fixes.  I appreciate it!
